### PR TITLE
Cow 22

### PR DESCRIPTION
**Pull request by tsoy-petr**
running tickets cow-22 and cow-23

### CODE REVIEW

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        super.updateDrawState(ds);
+        if (onclickListener != null) {
+            ds.setColor(onclickListener.mGetContext().getResources().getColor(R.color.middleBlue));
+        }
+    }
+
+    public interface StartSpannableOnclickListener {
+
+        public void start();
+        public Context mGetContext();
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java]: I think, better pass context in constructor, not thought interface. Context used only for getting color, instead context, you can pass required color

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    public void onClick(@NonNull View view) {
+        if (onclickListener != null) {
+            onclickListener.start();
+        }
+    }
+
+    @Override
+    public void updateDrawState(@NonNull TextPaint ds) {
+        super.updateDrawState(ds);
+        if (onclickListener != null) {
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java]: Not require check this.

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    private StartSpannableOnclickListener onclickListener;
+
+    public StartSpannable(StartSpannableOnclickListener onclickListener) {
+        this.onclickListener = onclickListener;
+    }
+
+    @Override
+    public void onClick(@NonNull View view) {
+        if (onclickListener != null) {
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java]: Not require check this,  you can add @NotNull annotation to constructor

File: app/src/main/res/layout/layout_keyboard.xml [hunk 4]
```diff
@@ -86,7 +86,7 @@
             <TextView
                 android:id="@+id/btn_start"
                 style="@style/AppTheme.KeyboardButton"
-                android:layout_height="40dp"
+                android:layout_height="27dp"
```
Comment (Kiolk) [inline app/src/main/res/layout/layout_keyboard.xml]: I think, There is no good idea decrease size of view, to make visible start word on specific device. Need other solution.

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/screens/rules/RulesFragment.java [hunk 5]
```diff
@@ -33,10 +34,14 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
         String coloredStartText = resources.getString(R.string.start_part);
         String howToPlayText = resources.getString(R.string.how_to_play_part);
         int length = text.length();
-        span.setSpan(new ForegroundColorSpan(view.getContext().getResources().getColor(R.color.middleBlue)), length - coloredStartText.length() - 1, length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
+
+        StartSpannable clickableSpan = new StartSpannable((StartSpannable.StartSpannableOnclickListener) getActivity());
+
+        span.setSpan(clickableSpan, length - coloredStartText.length() - 2, length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/screens/rules/RulesFragment.java]: Need to add logic not only activate the game, but and close the rule screen.

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/adapters/GameAdapter.java [hunk 6]
```diff
@@ -26,8 +27,12 @@
     private final int VIEW_TYPE_EMPTY = 0;
     private final int VIEW_TYPE_NORMAL = 1;
     private boolean isStartPressed = false;
+    private StartSpannable.StartSpannableOnclickListener onclickListener;
 
-    public GameAdapter() {
+    private GameAdapter(){}
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/adapters/GameAdapter.java]: Need remove empty adapter.

**Comment by Kiolk** [comment]
Good work, but need to fix some issues.

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java [hunk 7]
```diff
@@ -1,41 +1,43 @@
...
+        onclickListener.start();
     }
 
     @Override
     public void updateDrawState(@NonNull TextPaint ds) {
         super.updateDrawState(ds);
-        if (onclickListener != null) {
-            ds.setColor(onclickListener.mGetContext().getResources().getColor(R.color.middleBlue));
-        }
+//        ds.setColor(onclickListener.mGetContext().getResources().getColor(R.color.middleBlue));
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/views/StartSpannable.java]: Need remove commented code

File: app/src/main/java/com/github/kiolk/cowsandbulls/ui/screens/rules/RulesFragment.java [hunk 8]
```diff
@@ -30,14 +30,24 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
         TextView mRules = view.findViewById(R.id.tv_rules_description);
         Resources resources = view.getContext().getResources();
         String text = resources.getString(R.string.rules);
+        String coloredText = "";
+        String bottomText = "";
+        StartSpannable clickableSpan = new StartSpannable((StartSpannable.StartSpannableOnclickListener) getActivity(),
+                getContext().getResources().getColor(R.color.middleBlue));
+
+        if (clickableSpan.getOnclickListener().gameIsStarted()) {
```
Comment (Kiolk) [inline app/src/main/java/com/github/kiolk/cowsandbulls/ui/screens/rules/RulesFragment.java]: Sorry, I didn't clear explain. Need to add only logic for close rule fragment, not change word start to close. Please roll back changes depend on with it.

File: app/src/main/res/values-be/strings.xml [hunk 9]
```diff
@@ -16,9 +16,11 @@
         Адгадайце чатырохзначны лік, у якім лічбы не паўтараюцца.\n\n
         Пасля таго, як вы ўвядзеце меркаваны лік, вы атрымаеце інфармацыю пра колькасць:\n\n
         кароў - лічбы, якія ёсць у загаданым ліку, але знаходзяцца на іншай пазіцыі, \n\n
-        быкоў - лічбы, якія ёсць у загаданым ліку і знаходзяцца ў такой жа пазіцыі. \n\n
-        Каб распачаць гульню, націсніце \"Пачаць\"</string>
+        быкоў - лічбы, якія ёсць у загаданым ліку і знаходзяцца ў такой жа пазіцыі. \n\n</string>
     <string name="start_part">Пачаць</string>
+    <string name="close_part">Зачыніць</string>
```
Comment (Kiolk) [inline app/src/main/res/values-be/strings.xml]: Same. Need roll back to preview version.

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -30,9 +30,11 @@
         You need to guess a four-digit number with non-repeating digits.\n\n
         After entering the estimated number you will be given the answer where number of:\n\n
         cows - numbers are guessed without matching their positions, \n\n
-        bulls - numbers are guessed right up to the position. \n\n
-        To start click \"Start\"</string>
+        bulls - numbers are guessed right up to the position. \n\n</string>
```
Comment (Kiolk) [inline app/src/main/res/values/strings.xml]: Same. Need roll back to preview version.

**Review by Kiolk — COMMENTED** [review]
Need roll back some changes and add small fix

**Comment by tsoy-petr** [comment]
COW-22

**Pull request merged**